### PR TITLE
Fix type of self argument in cdef staticmethod declared in a pxd file

### PR DESCRIPTION
**Pull request by da-woods**
Fixes https://github.com/cython/cython/issues/3174

I've removed all assignment of `is_self_arg` out of the parser, since it can be overridden in too many ways so I don't think it's usefully done there. It's possible that the tests will show I've missed something important where its used (in which case I can undo the change)

I took the tests from @navytux's https://github.com/cython/cython/pull/3175/ (thanks!), so this PR will supercede that one

### CODE REVIEW

**Comment by navytux** [comment]
Thanks, @da-woods.

File: Cython/Compiler/Nodes.py [hunk 1]
```diff
@@ -901,8 +901,7 @@ def hdr_cname(self):
 
 
     def analyse(self, env, nonempty=0, is_self_arg=False):
-        if is_self_arg:
-            self.base_type.is_self_arg = self.is_self_arg = True
+        self.base_type.is_self_arg = self.is_self_arg = is_self_arg
```
Comment (scoder) [inline Cython/Compiler/Nodes.py:904]: Can we keep the `if` test? Given the default class attributes, it doesn't seem necessary to always set it (and that avoids an unnecessary attribute in the instance dicts).

File: Cython/Compiler/Parsing.py [hunk 2]
```diff
@@ -2477,13 +2477,11 @@ def p_positional_and_keyword_args(s, end_sy_set, templates = None):
         s.next()
     return positional_args, keyword_args
 
-def p_c_base_type(s, self_flag = 0, nonempty = 0, templates = None):
-    # If self_flag is true, this is the base type for the
-    # self argument of a C method of an extension type.
+def p_c_base_type(s, nonempty = 0, templates = None):
```
Comment (scoder) [inline Cython/Compiler/Parsing.py:2480]: Whenever we touch such code, it's a good time to clean it up along the way.
```suggestion
def p_c_base_type(s, nonempty=False, templates=None):
```
Note that you need to change `Parsing.pxd` as well (we should have a test job that exercises the `--cython-compile-more` build option – would also be interesting to see if it makes a performance dfference).

File: Cython/Compiler/Parsing.py [hunk 3]
```diff
@@ -2477,13 +2477,11 @@ def p_positional_and_keyword_args(s, end_sy_set, templates = None):
...
     if s.sy == '(':
         return p_c_complex_base_type(s, templates = templates)
     else:
-        return p_c_simple_base_type(s, self_flag, nonempty = nonempty, templates = templates)
+        return p_c_simple_base_type(s, nonempty = nonempty, templates = templates)
```
Comment (scoder) [inline Cython/Compiler/Parsing.py:2484]:
```suggestion
        return p_c_simple_base_type(s, nonempty=nonempty, templates=templates)
```

File: Cython/Compiler/Parsing.py [hunk 4]
```diff
@@ -2527,8 +2525,7 @@ def p_c_complex_base_type(s, templates = None):
     return type_node
 
 
-def p_c_simple_base_type(s, self_flag, nonempty, templates = None):
-    #print "p_c_simple_base_type: self_flag =", self_flag, nonempty
+def p_c_simple_base_type(s, nonempty, templates = None):
```
Comment (scoder) [inline Cython/Compiler/Parsing.py:2528]:
```suggestion
def p_c_simple_base_type(s, nonempty, templates=None):
```

File: Cython/Compiler/Parsing.py [hunk 5]
```diff
@@ -3052,7 +3049,7 @@ def p_c_arg_decl(s, ctx, in_pyfunc, cmethod_flag = 0, nonempty = 0,
             complex = 0, longness = 0,
             is_self_arg = cmethod_flag, templates = None)
     else:
-        base_type = p_c_base_type(s, cmethod_flag, nonempty = nonempty)
+        base_type = p_c_base_type(s, nonempty = nonempty)
```
Comment (scoder) [inline Cython/Compiler/Parsing.py:3052]:
```suggestion
        base_type = p_c_base_type(s, nonempty=nonempty)
```

**Comment by scoder** [comment]
Thanks @da-woods and @navytux !

File: Cython/Compiler/Parsing.py (hunk 2 above)
Comment (scoder) [reply Cython/Compiler/Parsing.py]: I'm working on adding CI jobs. Seems we've piled up some more differences in the past already.

**Pull request merged**